### PR TITLE
Snap to available properties

### PR DESCRIPTION
**Pull request by Munter**

### CODE REVIEW

File: lib/util/fonts/snapToAvailableFontProperties.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+};
+
+var INITIAL_VALUES = {
+    'font-family': undefined,
+    'font-stretch': 'normal',
+    'font-weight': 400,
+    'font-style': 'normal'
+};
+
+function ascendng(a, b) {
```
Comment (papandreou) [inline lib/util/fonts/snapToAvailableFontProperties.js]: ascending

File: lib/util/fonts/snapToAvailableFontProperties.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    });
+    propsToSnap = _.defaults({}, propsToSnap, INITIAL_VALUES);
+
+    // System font, we can't know about the full properties. Early exit
+    if (typeof propsToSnap['font-family'] === 'undefined') {
+        return propsToSnap;
+    }
+
+    // Match font-family first
+    var familymatches = _.filter(fontFaceDeclarations, { 'font-family': propsToSnap['font-family'] });
```
Comment (papandreou) [inline lib/util/fonts/snapToAvailableFontProperties.js]: Should probably be camel cased: `familyMatches`

File: lib/util/fonts/resolveFontWeight.js [hunk 3]
```diff
@@ -1,44 +1,33 @@
...
-            if (possibleWeight > 500) {
-                value = possibleWeight;
-            } else if (!value) {
-                value = possibleWeight;
-            }
-        });
-
-        return value;
-    }
+    return searchOrder[0];
```
Comment (papandreou) [inline lib/util/fonts/resolveFontWeight.js]: Looks like this code could be optimized quite a bit -- there's a bunch of intermediate arrays that get thrown away. OTOH it's very clear what's going on, so if it's not too hot we should probably leave it like this.

File: lib/util/fonts/snapToAvailableFontProperties.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    var stretchStartIndex = fontStretchValues.indexOf(propsToSnap['font-stretch']);
+    var stretchGroups = _.groupBy(familymatches, 'font-stretch');
+
+    var firstHalf, lastHalf, stretchSearchOrder;
+    var stretchMatches = [];
+
+    if (stretchStartIndex <= fontStretchValues.indexOf('normal')) {
+        // When value is 'normal' or lower, check denser values first, then less dense
+        firstHalf = fontStretchValues.slice(0, stretchStartIndex + 1);
+        lastHalf  = fontStretchValues.slice((stretchStartIndex + 1) - fontStretchValues.length);
```
Comment (papandreou) [inline lib/util/fonts/snapToAvailableFontProperties.js]: Unnecessary parentheses

File: lib/util/fonts/resolveFontWeight.js (hunk 3 above)
Comment (Munter) [inline lib/util/fonts/resolveFontWeight.js]: Yeah, this is pretty close to the described algorithm. But I think the description doesn't have the luxury of already having a limited subset of available weights, so it describes checking all weights.

I thought about early returns like `lastHalf[0] || firstHalf[firstHalf.length - 1]`, but it feel less clear from reading the code what the purpose it

File: lib/util/fonts/snapToAvailableFontProperties.js (hunk 4 above)
Comment (Munter) [inline lib/util/fonts/snapToAvailableFontProperties.js]: I'm using `stretchStartIndex + 1` in multiple places as a single construct, so I opted to keep these to make it easier to recognise

File: lib/util/fonts/snapToAvailableFontProperties.js (hunk 1 above)
Comment (Munter) [inline lib/util/fonts/snapToAvailableFontProperties.js]: Fixed

File: lib/util/fonts/snapToAvailableFontProperties.js (hunk 2 above)
Comment (Munter) [inline lib/util/fonts/snapToAvailableFontProperties.js]: Fixed

**Comment by papandreou** [comment]
Awesome! I'll take a look at CSS animations as soon as I get the chance.

**Pull request merged**